realm: Schedule realm deletion time while deactivating realm.

Ujjawal3 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #24677

Discussion: https://chat.zulip.org/#narrow/stream/101-design/topic/deactivate.20organization.20modal

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Screenshot from 2023-07-06 19-31-14](https://github.com/zulip/zulip/assets/99073049/7ae6e780-6a1a-42ca-97aa-8f59124b1786)
![Screenshot from 2023-07-06 19-28-36](https://github.com/zulip/zulip/assets/99073049/2d8643f5-1a52-44af-987c-037a39716a27)
![Screenshot from 2023-07-06 19-28-26](https://github.com/zulip/zulip/assets/99073049/5635595b-1082-474d-bc00-3aeb4543ff2b)
![Screenshot from 2023-07-06 19-28-09](https://github.com/zulip/zulip/assets/99073049/4dccd83f-e9d0-40f2-a8a1-728bf489dd16)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alexmv commented on zerver/tests/test_realm.py:
```diff
@@ -620,6 +622,69 @@ def test_deactivate_realm_by_non_owner(self) -> None:
+
+        # zephyr realm is set to delete after 6 minutes
+        current_time = timezone_now()
+        with mock.patch(
+            "zerver.actions.realm_settings.timezone_now",
+            return_value=current_time + timedelta(minutes=3),
+        ):
```
You should switch this to the `time_machine` library.

But it's not clear to me why you jump forward 3 minutes to ask to be deleted in 3 minutes, then jump back to now to check if it's _currently deleted_ (despite the deletion event being in the future at that time, which should be impossible), them jump forward 4 minutes, then back, then forward 6 minutes.

I think (with `time_machine`) this should be closer to:

- schedule the deletion in 3 minutes
- check that it's deactivated
- call delete_deactivated_realms, assert this doesn't do anything
- move forward 3 minutes
- call delete_deactivated_realms, assert `zephyr` is scrubbed
- move forward 7 more minutes
- call delete_deactivated_realms, assert `zulipinternal` is scrubbed

Ujjawal3 replied:
@alexmv I need few suggestions:
- **For asserting that calling delete_deactivated_realms does not do anything**
 Should I assert that there is no change in `RealmAuditLog` entries after calling the function? Or there is anything better we can do.
- **For asserting that a realm is scrubbed**
I could not think anything better other than asserting that count of Messages, Attachments and CustomProfile Fields associated with realm is zero.

Ujjawal3 commented on zerver/worker/queue_processors.py:
```diff
@@ -1153,6 +1154,13 @@ def failure_processor(event: Dict[str, Any]) -> None:
+            )
+            for realm in realms_to_scrub:
+                logger.info("Scrubbed realm %s", realm.id)
+                do_scrub_realm(realm, acting_user=None)
```
I am not sure about the right value of `acting_user` here.

alexmv replied:
`None` isn't wrong.  Alternately, we could look up who deactivated the realm last, and use that value.

Ujjawal3 commented on web/src/settings_config.ts:
```diff
@@ -513,6 +513,48 @@ export const expires_in_values = {
+    never: {
+        // Ideally we'd just store `null`, not the string `"null"`, but
+        // .val() will read null back as `""`.  Custom logic in
+        // get_common_realm_data converts this back to `null`
```
Yes, it was supposed to be in `do_deactivate_realm` I have made the changes

Ujjawal3 commented:
@alexmv I have addressed all your feedback and I think this PR is complete now except adding a backend test about which I have asked few questions in this [comment](https://github.com/zulip/zulip/pull/26109#discussion_r1245721339). So, you can take a look now.

Ujjawal3 commented:
@zulipbot add "buddy review"
@HemantUmre12 please have a look to this PR.

alexmv commented on puppet/zulip/files/cron.d/delete_deactivated_realms:
```diff
@@ -0,0 +1,6 @@
+USER=zulip
+
+# Cron job to clear data of deactivated realms.
+0 0 * * * zulip /home/zulip/deployments/current/manage.py delete_deactivated_realms
```
It's worth noting that this means that we may take up to 24 hours longer than expected to deleted a realm -- that is, if it expired at 00:00:01, then this cron job will next run nearly 24 hours later to delete it.

That's _probably_ fine -- increasing this frequency would reduce that slop, but 24h-less-immediate deletion seems fine.

alexmv commented on web/src/invite.js:
```diff
@@ -204,7 +210,13 @@ function valid_to(expires_in) {
 }
 
-function get_expiration_time_in_minutes() {
+export function get_expiration_time_in_minutes(
```
We should probably move this out to some other file, since having the realm deletion code import the invitation code is pretty clunky.

alexmv commented on web/src/invite.js:
```diff
@@ -204,7 +210,13 @@ function valid_to(expires_in) {
+    custom_expiration_time_unit,
+    custom_expiration_time_input,
+) {
+    if ($("#expires_in").val() !== "custom") {
```
We probably want to pass in `$("#expires_in")` to make the caller be explicit about the element it's acting on.

alexmv commented on web/src/settings_org.js:
```diff
@@ -702,29 +707,172 @@ export function discard_property_element_changes(elem, for_realm_default_setting
+    });
+
+    function deactivate_realm_modal_post_render() {
+        invite.set_custom_time_inputs_visibility(
+            custom_expiration_time_unit,
+            custom_expiration_time_input,
+        );
+
+        const expiration_time_in_minutes = invite.get_expiration_time_in_minutes(
+            custom_expiration_time_unit,
+            custom_expiration_time_input,
+        );
+        const valid_to_text = valid_to(expiration_time_in_minutes);
+        invite.set_expires_on_text(valid_to_text);
+
+        $("#expires_in").on("change", () => {
+            invite.set_custom_time_inputs_visibility(
+                custom_expiration_time_unit,
+                custom_expiration_time_input,
+            );
+
+            const expiration_time_in_minutes = invite.get_expiration_time_in_minutes(
+                custom_expiration_time_unit,
+                custom_expiration_time_input,
+            );
+            const valid_to_text = valid_to(expiration_time_in_minutes);
+            invite.set_expires_on_text(valid_to_text);
+        });
```
All of the `invite.` code in here needs to be moved out to someplace more generic -- this code has nothing to do with invitations.  It's a UI widget.

Ujjawal3 replied:
Hmm, was thinking to move the three functions `set_custom_time_inputs_visibility`, `get_expiration_time_in_minutes`,`set_expires_on_text` which are currently in `invite.js` to `settings_ui.js`. Is there any better place where these functions can be moved?

alexmv replied:
The frontend code is not my area of expertise, but that seems like a fine place for it to me.

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -1153,6 +1154,15 @@ def failure_processor(event: Dict[str, Any]) -> None:
+        elif event["type"] == "scrub_deactivated_realms":
+            realms_to_scrub = Realm.objects.filter(
+                deactivated=True,
+                scheduled_deletion_date__lte=timezone_now(),
+                scheduled_deletion_date__gt=timezone_now() - datetime.timedelta(days=1),
```
We should just limit to `scheduled_deletion_date__lte=timezone_now()` -- if the cron job gets skipped for any reason, then scrubbing the deactivated realms should scrub _all_ of the ones past the deletion date.

alexmv commented on zproject/default_settings.py:
```diff
@@ -566,3 +566,13 @@
 # More drafts, should they exist for some crazy reason, could be
 # fetched in a separate request.
 MAX_DRAFTS_IN_REGISTER_RESPONSE = 1000
+
+# Minimum number of days to be given in realm data
+# deletion schedule when deactivating a realm.
+# Give 0 if there is no minimum value.
+MIN_DEACTIVATED_REALM_DELETION_DAYS: int = 14
+
+# Maximum number of days that can be given in realm data
+# deletion schedule when deactivating a realm.
+# Give None if there is no maximum value.
+MAX_DEACTIVATED_REALM_DELETION_DAYS: Optional[int] = None
```
Dont' worry about the Zulip Cloud values -- we'll set those in `setting.py` on Zulip Cloud.

timabbott replied:
I improved these comments in `default_settings.py` and also deleted it from `prod_settings_template.py`; I don't think self-hosters will generally have a need to change these settings, and they're not easy to document clearly for that audience.

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -1153,6 +1154,15 @@ def failure_processor(event: Dict[str, Any]) -> None:
             )
             user_profile = get_user_profile_by_id(event["user_profile_id"])
             reactivate_user_if_soft_deactivated(user_profile)
+        elif event["type"] == "scrub_deactivated_realms":
+            realms_to_scrub = Realm.objects.filter(
```
I think we should have the event be for just _one_ `realm_id`, which is provided in the `event` -- and move this `Realm.objects.filter()` into the management command.  That "chunks" the work into smaller events.

alexmv commented on zerver/actions/realm_settings.py:
```diff
@@ -527,3 +546,10 @@ def do_send_realm_reactivation_email(realm: Realm, *, acting_user: Optional[User
     )
+
+
+def delete_deactivated_realms() -> None:
+    deferred_work_event = {
+        "type": "scrub_deactivated_realms",
+    }
+    queue_json_publish("deferred_work", deferred_work_event)
```
As I note below, I think this function should take a `realm: Realm` and enqueue an event for just one `realm_id`.  The reason for this is assume that 20 realms all deactivate at once, with immediate deletion -- that will end up with 20 "delete all expired realms" events, which is _safe_ (as long as we only have one queue worker) but some unnecessary queries.

We should also name this function `scrub_deactivated_realm`, for accuracy.

Ujjawal3 replied:
Should we add a check whether realm should be scrubbed now or not inside the function like this 
```
if realm.scheduled_deletion_date is None or realm.scheduled_deletion_date > timezone_now():
    return
```
I was adding this so that we can write test for this function or should I write test for command in `test_management_commands`.

alexmv replied:
We _can_ do that here, but correctness only says that we need to do it when we're in the queue worker, right before we're doing the work.  If we check it here, any number of things could happen between now, when we enqueue the work, and some arbitrary amount of time later when the worker actually gets to it.

alexmv commented on puppet/zulip/manifests/app_frontend_once.pp:
```diff
@@ -51,6 +51,14 @@
     source => 'puppet:///modules/zulip/cron.d/soft-deactivate-users',
   }
 
+  file { '/etc/cron.d/delete_deactivated_realms':
+    ensure => file,
+    owner  => 'root',
+    group  => 'root',
+    mode   => '0644',
+    source => 'puppet:///modules/zulip/cron.d/delete_deactivated_realms',
```
We should probably name this `scrub_deactivated_realms` for accuracy.

alexmv commented on zerver/management/commands/delete_deactivated_realms.py:
```diff
@@ -0,0 +1,11 @@
+from typing import Any
+
+from zerver.actions.realm_settings import delete_deactivated_realms
+from zerver.lib.management import ZulipBaseCommand
+
+
+class Command(ZulipBaseCommand):
+    help = """Clears data of deactivated realms."""
+
+    def handle(self, *args: Any, **options: Any) -> None:
+        delete_deactivated_realms()
```
As noted elsewhere, `scrub_deactivated_realms` is a better name than `delete_deactivated_realms`

alexmv commented on zproject/prod_settings_template.py:
```diff
@@ -837,3 +837,15 @@
 ## Directory containing Markdown files for the server's policies.
 # POLICIES_DIRECTORY = "/etc/zulip/policies/"
+
+# Schedule deactivated realm data deletion
+
+# Minimum number of days to be given in realm data
+# deletion schedule when deactivating a realm.
+# Give 0 if there is no minimum value.
+# MIN_DEACTIVATED_REALM_DELETION_DAYS = 14
+
+# Maximum number of days that can be given in realm data
+# deletion schedule when deactivating a realm.
+# Give None if there is no maximum value.
+# MAX_DEACTIVATED_REALM_DELETION_DAYS= None
```
We put two `#` before comments, and we should make clear the _reason_ you would (or would not) set None or 0 for these, not just what they entail.

```suggestion
### Data retention settings
## Minimum number of days a realm administrator can
## choose to have data deleted after.  Set to 0 to
## allow realm administrators to deactivate and
## immediately irrevocably delete all data.
# MIN_DEACTIVATED_REALM_DELETION_DAYS = 14

## Maximum number of days a realm administrator can
## choose to have data deleted after.  Set to None
## to allow realm administrators to never delete data.
# MAX_DEACTIVATED_REALM_DELETION_DAYS= None

Ujjawal3 commented:
@alexmv I have locally done changes for most of the comments. I have asked a couple of minor questions in inline commands would be great if you can answer them. After this PR would be ready for review.

Ujjawal3 commented:
@alexmv I have adressed all your commemnts. Please have a look.

alexmv commented on zerver/views/realm.py:
```diff
@@ -352,9 +354,32 @@ def update_realm(
+        realm_deletion_in_minutes is None
+        and settings.MAX_DEACTIVATED_REALM_DELETION_DAYS is not None
+    ):
+        raise JsonableError(_("Given data deletion time is not allowed."))
```
We might want a more specific error here -- something about `Realms must provide a realm data deletion timeline of {__} or fewer days`

alexmv commented on zerver/views/realm.py:
```diff
@@ -352,9 +354,32 @@ def update_realm(
+            and realm_deletion_in_minutes > settings.MAX_DEACTIVATED_REALM_DELETION_DAYS * 24 * 60
+        )
+    ):
+        raise JsonableError(_("Given data deletion time is not allowed."))
```
Similarly, we can be helpful and provide something like `Given data deletion time is not allowed; it must be between {__} and {___} days`

alexmv reviewed (commented):
You should get someone else to review the frontend changes -- they still feel a little oddly-factored to me, but I'm not familiar enough with the code and conventions to have specific suggestions there.

My only remaining comments are pretty small nitpicks; this is shaping up well!

Ujjawal3 commented:
@sahil839 can you take a look at frontend changes in this PR?

sahil839 commented on web/templates/confirm_dialog/confirm_deactivate_realm.hbs:
```diff
@@ -1,2 +1,26 @@
+        <p class="allowed_duration_message">{{allowed_duration_message}}</p>
+        <label for="expires_in">{{t "Custom time" }}</label>
+        <input type="text" autocomplete="off" name="custom-expiration-time" id="custom-expiration-time-input" class="custom-expiration-time inline-block" value="" maxlength="3"/>
+        <select class="custom-expiration-time bootstrap-focus-style" id="custom-expiration-time-unit">
```
Using expiration time here feels weird as we use "expire" for invites and we instead use "delete" for realms. Would be good to rename these and also refactor the common functions to receive the element as parameter instead of using `$("#expires_in")`.

sahil839 replied:
There can be two options - either you can pass the containing div of all these elements or you can just pass the main select element `$("#expires_in")` and then to get the custom element you can do something like `$("#expires_in").parent().find({some-common-class-used-for-custom-inputs})`. 

think we use the second approach in custom limit organization settings, so would be good to use same here as well. You can re-check this once to confirm what we do for organization settings.

sahil839 commented on web/templates/confirm_dialog/confirm_deactivate_realm.hbs:
```diff
@@ -1,2 +1,26 @@
+    <div id="custom-deactivate-realm-expiration-time" class="dependent-settings-block custom-time-input-container">
+        <p class="allowed_duration_message">{{allowed_duration_message}}</p>
+        <label for="expires_in">{{t "Custom time" }}</label>
+        <input type="text" autocomplete="off" name="custom-expiration-time" id="custom-expiration-time-input" class="custom-expiration-time inline-block" value="" maxlength="3"/>
```
I think you need to add `modal_text_input` class here.

sahil839 commented on web/src/settings_config.ts:
```diff
@@ -513,6 +513,48 @@ export const expires_in_values = {
+        description: $t({defaultMessage: "Immediately"}),
+        default: false,
+    },
+    fourteenDays: {
```
I think we can use snake_case variables here instead of camelCase variables. I think we can update the ones in `expires_in_values` as well, since all others seem to use snake_case.

sahil839 commented on web/src/invite.js:
```diff
@@ -294,18 +252,18 @@ function open_invite_user_modal(e) {
         );
         const valid_to_text = valid_to(expiration_time_in_minutes);
-        set_expires_on_text(valid_to_text);
+        settings_ui.set_expires_on_text(valid_to_text);
```
I am not sure whether moving these to `settings_ui` is correct since I guess we do not consider invite modal as part of settings UI as per the other classes used for it. But I am not sure what would be the better place though, so we can keep it as it is for now.

timabbott replied:
I think that's probably OK -- it's a generic helper for managing our settings-style inputs; I think of the "invitation" form as a settings modal that happens to not be inside the main settings UI...

sahil839 reviewed (commented):
Posted a few initial comments about naming and refactoring. Have not reviewed completely yet, would do so later today.

Ujjawal3 commented:
@sahil839 I have updated the PR and have also resolved all the merge conflicts. You can take a look now.

Ujjawal3 commented:
@sahil839 I have updated the PR. You can take a look now.

sahil839 commented on web/src/settings_org.js:
```diff
@@ -702,29 +706,191 @@ export function discard_property_element_changes(elem, for_realm_default_setting
+            }
+        });
+
+        $(".custom-time-input-container").on(
```
Here also would be good to use the ID of this element instead of this class.

sahil839 commented on web/src/settings_org.js:
```diff
@@ -702,29 +706,191 @@ export function discard_property_element_changes(elem, for_realm_default_setting
+    }
+
+    if (!time_valid) {
+        return $t({defaultMessage: "Never deleted"});
```
Can we modify this text to be "Data would never be deleted"?

alexmv replied:
`Data will not be automatically deleted` is perhaps more accurate?

Ujjawal3 replied:
I have modified the text to use `Data will not be automatically deleted`.

sahil839 reviewed (commented):
Posted a few comments. I am not sure about the class and function names used for the text mentioning the expiration and deletion time. But cannot think of something better currently.

Ujjawal3 commented:
@sahil839 I have updated the PR and have addressed the comments other than renaming. I could not think of better variable names.

sahil839 commented on web/styles/zulip.css:
```diff
@@ -2413,6 +2413,7 @@ textarea.invitee_emails {
 }
 
 select.input-time-choices,
+select.deactivated-realm-expires-in,
```
This class is not used anywhere.

sahil839 commented on web/styles/zulip.css:
```diff
@@ -2426,6 +2427,10 @@ select.invite-as {
 
+#deactivate-realm-user-modal .allowed_duration_message {
+    font-style: italic;
+}
```
Would be better to have this CSS in `settings.css` since that is where we have the CSS for other modals opened from settings overlay.

sahil839 commented on web/styles/zulip.css:
```diff
@@ -2481,7 +2481,7 @@ select.invite-as {
 }
 
-#custom-expiration-time-unit {
+.custom-input-time-unit {
```
This change makes the width of unit dropdown also 220px.

Ujjawal3 replied:
Hmm, earlier it was making width `220px` but now after using `modal_select` class there is now change in width.

sahil839 commented on web/styles/zulip.css:
```diff
@@ -2407,13 +2407,13 @@ textarea.invitee_emails {
-select.invite-expires-in,
-select.custom-expiration-time,
+select.input-time-choices,
+select.custom-input-time-unit,
```
I am not sure whether we need to have this change. For other modals opened from settings we use `modal_select` class to apply this CSS.

We could also try changing the invite modal as well to use that class. I guess at the time we added `modal_select` class, invite UI used to be an overlay and not modal.

Ujjawal3 replied:
I have removed this complete css and have added `modal_select` class to all select elements.

sahil839 commented on web/templates/confirm_dialog/confirm_deactivate_realm.hbs:
```diff
@@ -1,2 +1,26 @@
+{{/if}}
+<div class="input-group">
+    <label for="deletes_in">{{t "Automatically delete all data after" }}</label>
+    <select id="deletes_in" class="input-time-choices bootstrap-focus-style">
```
Here also we should add `modal_select` class.

sahil839 reviewed (commented):
Posted a few more comments which I missed earlier. They are mostly about the CSS changes.

Ujjawal3 commented:
@sahil839 I have done the changes as per your comments. Please have a look.

sahil839 commented on web/src/settings_org.js:
```diff
@@ -702,29 +706,191 @@ export function discard_property_element_changes(elem, for_realm_default_setting
+            );
+        }
+        const valid_to_text = valid_to(deletion_time_in_minutes);
+        settings_ui.set_valid_till_text($("#deletes_in"), valid_to_text);
```
This code is duplicated and is used two times in invite code apart form being used twice here. Would have been good to have a common function doing this but `valid_to` is different for invite case, so I am not sure. We can probably pass `valid_to` as an argument to the common function. Let's wait for more feedback on this.

timabbott replied:
Probably best to leave this as a follow-up; I agree it's messy but I'm not sure what the best fix is.

sahil839 commented:
LGTM. @timabbott ready for your review. I am not sure about the naming of couple of functions like `valid_till_text` and the corresponding classes and also posted a comment about refactoring above.

alexmv commented on zerver/views/realm.py:
```diff
@@ -351,9 +353,39 @@ def update_realm(
+    ):
+        raise JsonableError(
+            _(
+                "Given data deletion time is not allowed;it must be between {min_allowed_days} and {max_allowed_days} days."
```
```suggestion
                "Given data deletion time is not allowed; it must be between {min_allowed_days} and {max_allowed_days} days."
```

alexmv commented on zerver/views/realm.py:
```diff
@@ -351,9 +353,39 @@ def update_realm(
+        realm_deletion_in_minutes < settings.MIN_DEACTIVATED_REALM_DELETION_DAYS * 24 * 60
+    ):
+        raise JsonableError(
+            _("Data deletion timeline must be of {min_allowed_days} or more days.").format(
```
```suggestion
            _("Data deletion time must be {min_allowed_days} or more days.").format(
```

alexmv commented on zproject/prod_settings_template.py:
```diff
@@ -673,6 +673,21 @@
+## Maximum number of days a realm administrator can
+## choose to have data deleted after.  Set to None
+## to allow realm administrators to never delete data.
+# MAX_DEACTIVATED_REALM_DELETION_DAYS= None
```
```suggestion
# MAX_DEACTIVATED_REALM_DELETION_DAYS = None
```

alexmv reviewed (commented):
Glad this is close to the finish line!  I added a couple tiny nits I saw in passing.

One other thought occurs to me -- does `scheduled_deletion_date` get unset when a realm is scrubbed?  Because if not, then we'll re-enqueue it the next time the cron job runs, and re-scrub an ever increasing list of realms.

Can you add a test for this?

Ujjawal3 commented:
> Glad this is close to the finish line! I added a couple tiny nits I saw in passing.
> 
> One other thought occurs to me -- does `scheduled_deletion_date` get unset when a realm is scrubbed? Because if not, then we'll re-enqueue it the next time the cron job runs, and re-scrub an ever increasing list of realms.
> 
> Can you add a test for this?

Yes, Earlier I missed the part of unsetting when a realm is scrubbed. Now, I have made changes and `scheduled_deletion_time` is made to `null` once a realm has been scrubbed, so the realm will not be enqueued next time. I have not added a separate test for this instead have tested this in `test_scrub_deactivated_realms`.

timabbott commented:
@Ujjawal3 just a ping that it'd be great to get this finished so that we can integrate all your work on it!

Ujjawal3 commented:
@timabbott this PR is completed. I have rebased the PR to resolve the merge conflicts. You can integrate it now.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
 ## Changes in Zulip 8.0
 
+**Feature level 205**
+* [`POST /register`](/api/register-queue): Add `min_deactivated_realm_deletion_days`
```
The missing blank line here results in this page being misrendered; it's best to always test documentation pages because this kind of typographical issue happens all the time if you don't.

timabbott commented on zerver/worker/queue_processors.py:
```diff
@@ -1139,6 +1140,17 @@ def failure_processor(event: Dict[str, Any]) -> None:
+                do_scrub_realm(realm_to_scrub, acting_user=None)
+                realm_to_scrub.scheduled_deletion_date = None
+                realm_to_scrub.save()
+                logger.info("Scrubbed realm %s", realm_to_scrub.id)
```
What's the thinking on this going into a queue? I don't see a reason the cron job couldn't process things directly.

Ujjawal3 replied:
Yeah, it was done for the `immediate` option to work.

timabbott commented on zerver/views/realm.py:
```diff
@@ -351,9 +353,39 @@ def update_realm(
+    request: HttpRequest,
+    user: UserProfile,
+    realm_deletion_in_minutes: Optional[int] = REQ(
+        json_validator=check_none_or(check_int), default=None
```
This API change is missing an API documentation note; we don't have docs for this endpoint so we only need something in the API changelog. I'll fix.

Also renamed to `deletion_delay_minutes` ... though we may want to change the units to days.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11226,6 +11226,25 @@ paths:
 
                           **Changes**: New in Zulip 4.0 (feature level 53). Previously,
                           this property always had a value of 10000.
+                      min_deactivated_realm_deletion_days:
```
I added a `server_` prefix to these API variable names.

timabbott commented on web/templates/confirm_dialog/confirm_deactivate_realm.hbs:
```diff
@@ -1,2 +1,26 @@
+    <div id="custom-realm-deletion-time" class="dependent-settings-block custom-time-input-container">
+        <p class="allowed_duration_message">{{allowed_duration_message}}</p>
+        <label for="deletes_in">{{t "Custom time" }}</label>
+        <input type="text" autocomplete="off" name="custom-deletion-time" id="custom-deletion-time-input" class="custom-input-time-value inline-block modal_text_input" value="" maxlength="3"/>
```
The maxlength of 3 here makes sense for invitations, which fundamentally we should be OK with expiring fairly quickly, but doesn't feel right for this input -- changed to 4.

timabbott commented on web/src/settings_org.js:
```diff
@@ -704,29 +708,191 @@ export function discard_property_element_changes(elem, for_realm_default_setting
+        return $t(
+            {
+                defaultMessage:
+                    "Enter a value between {minimum_allowed_days} and {maximum_allowed_days} days",
```
I don't love having two new strings here; it seems like it'd be better to just use the same strings we have in the server, which I've changed to be, for example, `"{field_name} must be at most {max_allowed_days} days."`

Ujjawal3 replied:
I think we should include values of both `max_allowed_days` and `min_allowed_days` in this string. This string is not an error message, it will be shown when user is entering a custom time like in screenshot below. 
![Screenshot from 2023-07-06 19-31-14](https://github.com/zulip/zulip/assets/99073049/7ae6e780-6a1a-42ca-97aa-8f59124b1786)

timabbott replied:
The problem with that approach is if one of those two values is `None`, the strings get very awkward -- "Enter a value between None and 100 days" looks bad.

timabbott commented on web/src/settings_config.ts:
```diff
@@ -513,6 +513,48 @@ export const expires_in_values = {
+        value: 365 * 24 * 60,
+        description: $t({defaultMessage: "1 Year"}),
+        default: false,
+    },
```
Added a "2 years" option and fixed the capitalization for "1 year" to match the other options.

timabbott commented:
@Ujjawal3 thanks for all the work on this; I've pushed back a version with several edits detailed above. A few requests not remarked on above:
* I think the API should be switched to `_days`, not `_minutes` -- this was discussed in the #design thread. (You may find you want the JS code to convert units just before talking to the API so that you can use the reusable web app component built in terms of minutes)
* I think we should make `scrub_deactivated_realm` be the name of a function that is the block of code that you currently have in the `deferred_work` queue processor for doing the scrubbing; and the cron job can call that directly (without using the queue processor), such that only the `immediately` option would use the queue processor, and the queue processor would have very little code.

I have a few product decisions to settle that might result in a few other small changes, but I think those are the main things I need from you to merge this.

Ujjawal3 commented:
@timabbott I have done the changes. You can take a look now.

timabbott commented:
@Ujjawal3 cool, let me know the resolution to https://github.com/zulip/zulip/pull/26109#discussion_r1308095044; also you probably need to renumber the API changelog stuff again as I think I just merged a little PR that bumps to 206 today.

Ujjawal3 commented:
I have rebased the PR to bump the feature level entry.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -514,3 +537,14 @@ def do_send_realm_reactivation_email(realm: Realm, *, acting_user: Optional[User
+    ):
+        do_scrub_realm(realm_to_scrub, acting_user=None)
+        realm_to_scrub.scheduled_deletion_date = None
+        realm_to_scrub.save()
```
I think `do_scrub_realm` should be responsible for clearing `scheduled_deletion_date`; then when we add an `@transaction.atomic` to that function, this step can't be missed.

Ujjawal3 replied:
`do_scrub_realm` is called from a few other places as well. Using `git grep` outputs the following 
```
analytics/tests/test_support_views.py:700:        with mock.patch("analytics.views.support.do_scrub_realm") as m:
analytics/tests/test_support_views.py:707:        with mock.patch("analytics.views.support.do_scrub_realm") as m:
analytics/views/support.py:27:    do_scrub_realm,
analytics/views/support.py:281:            do_scrub_realm(realm, acting_user=acting_user)
zerver/actions/realm_settings.py:376:def do_scrub_realm(realm: Realm, *, acting_user: Optional[UserProfile]) -> None:
zerver/actions/realm_settings.py:547:        do_scrub_realm(realm_to_scrub, acting_user=None)
zerver/management/commands/scrub_realm.py:6:from zerver.actions.realm_settings import do_scrub_realm
zerver/management/commands/scrub_realm.py:24:        do_scrub_realm(realm, acting_user=None)
``` 
Although I don't think there will be any issue if we clear `scheduled_deletion_date` inside the function but I think we can add `@transaction.actomic` to `scrub_deactivated_realms` itself. Curently I have not done any changes let me know which one is better.

timabbott replied:
I think `do_scrub_realm` should always clear the `scheduled_deletion_date`, and do so in a transaction -- fundamentally, we're deleting the realm, and that's a bit of data that no longer has meaning once the realm has been deleted.

Ujjawal3 replied:
Ok, I have done the changes.

timabbott commented on web/src/settings_org.js:
```diff
@@ -704,29 +708,193 @@ export function discard_property_element_changes(elem, for_realm_default_setting
+                    custom_deletion_time_input,
+                ),
+            );
+            deletes_in_days = deletes_in_minutes / (60 * 24);
```
What happens if the custom time is not an integer number of days here -- does this round or send a floating point value?

Ujjawal3 replied:
No, the value is not rounded. Instead floating number is sent to API and API is returning a json error `deletion_delay_days is not an integer`. I don't think we need to make any changes in this behaviour. Let me know if any thing needs to be changed.

sahil839 replied:
I think if we only want to allow the users to set value in days, then we should not provide other options in the UI. It does not make sense to give users the option of choosing time unit in minutes and then expect them to give the input such that it can be converted to integer number of days.

Ujjawal3 replied:
We are not allowing users to give time in terms of minutes. User can only give time in terms of number of days or weeks. The question above was what If a user gives a value like 3.2 days.

sahil839 replied:
Ok. So we can prevent that in UI itself I think. We can just disable the submit button of modal in that case I guess.

timabbott replied:
Yeah that seems reasonable; @Ujjawal3 have you done something like that, or is this item still open? I think it could reasonably be left as a follow-up issue because I don't think anyone is in practice going to enter `3.2` anyway.

Ujjawal3 replied:
No, I have not done any work for it.

laurynmm commented:
Closing in favor of https://github.com/zulip/zulip/pull/31957, which builds on the final changes that remain from this pull request.
